File: 黑马视频练习/MyWeiBo/MyWeiBo/Class/Guide(引导页)/Controller/HJGuideController.m
//
//  HJGuideController.m
//  MyWeiBo
//
//  Created by M-coppco on 16/5/9.
//  Copyright © 2016年 XHJ. All rights reserved.
//

#import "HJGuideController.h"
#define kGuideNum 4
#define kGuideIndentify @"GuideCell"
#import "HJTabBarController.h"

@interface HJGuideController ()<UICollectionViewDataSource, UICollectionViewDelegate>
@property (nonatomic, strong)UICollectionView  *collection;
@property (nonatomic, strong)UICollectionViewFlowLayout  *layout;
@property (nonatomic, strong)NSMutableArray  *images;//图片
@property (nonatomic, strong)UIPageControl  *pageControl;
@property (nonatomic, strong)UIButton  *beginB;  //开始微博
@property (nonatomic, strong)UIButton  *shareB; //分享按钮
@end

@implementation HJGuideController
- (UIButton *)beginB {
    if (!_beginB) {
        _beginB = [UIButton buttonWithType:(UIButtonTypeCustom)];
        [_beginB setTitle:@"开始微博" forState:(UIControlStateNormal)];
        [_beginB setTitleColor:[UIColor blackColor] forState:(UIControlStateNormal)];
        [_beginB setBackgroundImage:[UIImage imageNamed:@"new_feature_finish_button_highlighted"] forState:(UIControlStateHighlighted)];
        [_beginB setBackgroundImage:[UIImage imageNamed:@"new_feature_finish_button"] forState:(UIControlStateNormal)];
        [_beginB addTarget:self action:@selector(shareB:) forControlEvents:(UIControlEventTouchUpInside)];
    }
    return _beginB;
}
- (UIButton *)shareB {
    if (!_shareB) {

        _shareB = [UIButton buttonWithType:(UIButtonTypeCustom)];
        _shareB.adjustsImageWhenHighlighted = NO; //长按变灰效果去掉
        [_shareB setTitle:@"分享给大家" forState:(UIControlStateNormal)];
        [_shareB setTitleColor:[UIColor blackColor] forState:(UIControlStateNormal)];
        [_shareB setImage:[UIImage imageNamed:@"new_feature_share_false"] forState:(UIControlStateNormal)];
        [_shareB addTarget:self action:@selector(shareB:) forControlEvents:(UIControlEventTouchUpInside)];
        [_shareB setImage:[UIImage imageNamed:@"new_feature_share_true"] forState:(UIControlStateSelected)];
        _shareB.titleEdgeInsets = UIEdgeInsetsMake(0, 5, 0, 0);
        _shareB.imageEdgeInsets = UIEdgeInsetsMake(0, 0, 0, 5);
    }
    return _shareB;
}
- (void)shareB:(UIButton *)btn {
    if (btn == self.shareB) {
        btn.selected = !btn.selected;
    } else {
        NSUserDefaults *userD = [NSUserDefaults standardUserDefaults];
        [userD setObject:[HJTool getCFBundleVersonFromNSBundle] forKey:@"CFBundleVersion"];
        [userD synchronize];
        [HJTool animationGradualForView:self.collection type:(AnimateTypeBig) isRotateFow:YES delegate:self];
        
    }
}
- (void)viewDidLoad {
    [super viewDidLoad];
    [self getImages];
    [self configCollectionViewAndPageControl];
   
}
- (void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag {
    //跳转
    getAppWindow().rootViewController = [[HJTabBarController alloc] init];
    if (ISIOS_8_0) {
        //注册通知
        UIUserNotificationSettings* settings=[UIUserNotificationSettings settingsForTypes:UIUserNotificationTypeAlert|UIUserNotificationTypeBadge|UIUserNotificationTypeSound categories:nil];
        [[UIApplication sharedApplication] registerUserNotificationSettings:settings];
    } else {
        UIRemoteNotificationType type = UIRemoteNotificationTypeAlert | UIRemoteNotificationTypeBadge | UIRemoteNotificationTypeSound;
        [[UIApplication sharedApplication] registerForRemoteNotificationTypes:type];
    }
}
- (void)getImages {
    self.images = [NSMutableArray array];
    for (int  i = 0;  i < kGuideNum; i++) {
        NSString *str = [NSString stringWithFormat:@"new_feature_%d", i + 1];
        [self.images addObject:str];
    }
}
- (void)configCollectionViewAndPageControl {
    self.layout = [[UICollectionViewFlowLayout alloc] init];
    self.layout.itemSize = CGSizeMake(KMainScreenWidth, KMainScreenHeight);
    self.layout.scrollDirection = UICollectionViewScrollDirectionHorizontal; //水平
    self.layout.minimumInteritemSpacing = 0;
    self.layout.minimumLineSpacing = 0;
    
    self.collection = [[UICollectionView alloc] initWithFrame:KMainScreenBounds collectionViewLayout:self.layout];
    self.collection.pagingEnabled = YES;
    self.collection.showsHorizontalScrollIndicator = NO;
    self.collection.bounces = NO;  //边界回弹为NO
    
    //注册单元格
    [self.collection registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:kGuideIndentify];
    
    self.collection.delegate = self;
    self.collection.dataSource = self;
    [self.view addSubview:self.collection];
    
    self.pageControl = [[UIPageControl alloc] init];
    self.pageControl.numberOfPages = kGuideNum;
    self.pageControl.currentPage = 0;
    self.pageControl.currentPageIndicatorTintColor = [UIColor redColor];
    self.pageControl.pageIndicatorTintColor = [UIColor greenColor];
    [self.view addSubview:self.pageControl];
    
    [self.pageControl mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.mas_equalTo(self.view).offset(-20);
        make.left.right.mas_equalTo(self.view);
    }];
}
- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    int page = self.collection.contentOffset.x / self.collection.width;
    self.pageControl.currentPage = page;
}
#pragma mark - UICollectionViewDataSource和UICollectionViewDelegate
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return kGuideNum;
}

// The cell that is returned must be retrieved from a call to -dequeueReusableCellWithReuseIdentifier:forIndexPath:
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:kGuideIndentify forIndexPath:indexPath];
    UIImage *image = [UIImage imageNamed:self.images[indexPath.item]];
    UIImageView *imageV = [cell viewWithTag:9999];
    if (!imageV) {
        UIImageView *imageV = [[UIImageView alloc] initWithFrame:KMainScreenBounds];
        imageV.image = image;
        imageV.tag = 9999;
        imageV.userInteractionEnabled = YES;
        [cell.contentView addSubview:imageV];
    }
    imageV.image = image;

    if (indexPath.row == kGuideNum - 1) {
        [imageV addSubview:self.beginB];
        [imageV addSubview:self.shareB];
        self.shareB.hidden = NO;
        self.beginB.hidden = NO;
        [self.shareB mas_makeConstraints:^(MASConstraintMaker *make) {
            make.centerX.mas_equalTo(imageV);
            make.centerY.mas_equalTo(imageV).offset(50);
            make.left.right.mas_equalTo(imageV);
            make.height.mas_equalTo(30);
        }];
        [self.beginB mas_makeConstraints:^(MASConstraintMaker *make) {
            make.centerX.mas_equalTo(imageV);
            make.centerY.mas_equalTo(self.shareB).offset(40);
            make.width.mas_equalTo(100);
            make.height.mas_equalTo(30);
        }];
    } else if (indexPath.row % 2 == (kGuideNum % 2) ? 0 : 1){
        //判断  和最后一页相隔2个的都让它隐藏
        self.shareB.hidden = YES;
        self.beginB.hidden = YES;
    }
    return cell;
}
//隐藏状态栏
- (BOOL)prefersStatusBarHidden {
    return YES;
}
/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: 小码哥大神5期练习/6、Swift版本-百思不得姐/Best Not Elder Sister/Best Not Elder Sister/Class/Base/Controller/HJGuideController.swift
//
//  HJGuideController.swift
//  Best Not Elder Sister
//
//  Created by coco on 16/7/22.
//  Copyright © 2016年 M-coppco. All rights reserved.
//

import UIKit

private let reuseIdentifier = "Cell"

class HJGuideController: UICollectionViewController {

    /**点击跳过或者进入的闭包*/
    var skipClosure:(() ->Void)?
    
    private var images: [String] = [String]()
    private let number: Int = 4
    private let layout: UICollectionViewFlowLayout!
    private lazy var skipButton: UIButton = {
        let button = UIButton(type: UIButtonType.Custom)
        button.frame = CGRectMake(kHJMainScreenWidth - 50 - 20, 20, 60, 30)
        button.setTitle("跳过", forState: .Normal)
        button.backgroundColor = UIColor.brownColor().colorWithAlphaComponent(0.5)
        //按钮的遮罩
        let bezierPath = UIBezierPath(roundedRect: button.bounds, byRoundingCorners: [.TopRight, .BottomRight], cornerRadii: CGSizeMake(15, 15))
        let mask = CAShapeLayer();
        mask.frame = button.bounds
        mask.path = bezierPath.CGPath
        button.layer.mask = mask
        return button
    }()
    
    private lazy var enterButton:UIButton = {
        let button = UIButton(type: UIButtonType.Custom)
        button.setTitle("进入逗比的世界", forState: .Normal)
        button.setTitleColor(UIColor.whiteColor(), forState: .Normal)
        button.titleLabel?.font = UIFont.systemFontOfSize(15)
        button.backgroundColor = UIColor.grayColor().colorWithAlphaComponent(0.7)
        button.layer.cornerRadius = 15
        button.hidden = true
        return button
    }()
    
    
    private lazy var pageControl: UIPageControl = {
        let page = UIPageControl(frame: CGRectMake(self.view.frame.size.width / 2 - 50, self.view.frame.size.height - 40,100,30))
        return page
    }()
    
    /**默认初始化方法*/
    init(closure: () -> Void) {
        layout = UICollectionViewFlowLayout()
        layout.scrollDirection = .Horizontal
        layout.itemSize = kHJMainScreenBounds.size
        layout.minimumInteritemSpacing = 0
        layout.minimumLineSpacing = 0
        layout.sectionInset = UIEdgeInsetsZero
        for i in 1...number {
            if is_iPhone4 {
                images.append(String(format: "320*480-%d.jpg", i))
            }else if is_iPhone4s {
                images.append(String(format: "640*960-%d.jpg", i))
            }else if is_iPhone5 {
                 images.append(String(format: "640*1136-%d.jpg", i))
            }else if is_iPhone6 {
                 images.append(String(format: "750*1334-%d.jpg", i))
            }else if is_iPhone6p {
                 images.append(String(format: "1242*2208-%d.jpg", i))
            } else {
                images.append(String(format: "1242*2208-%d.jpg", i))
            }
        }
        skipClosure = closure
        super.init(collectionViewLayout: layout)
    }

     required init?(coder aDecoder: NSCoder) {
         fatalError("init(coder:) has not been implemented")
     }

    deinit {
        HJLog("\(self.classForCoder)","释放了")
    }
    
    /**使用存放图片名称的数组初始化, 跳过闭包*/
    init(imageArray:[String], closure:() ->Void) {
        layout = UICollectionViewFlowLayout()
        layout.scrollDirection = .Horizontal
        layout.itemSize = kHJMainScreenBounds.size
        layout.minimumInteritemSpacing = 0
        layout.minimumLineSpacing = 0
        layout.sectionInset = UIEdgeInsetsZero
        images = imageArray
        self.skipClosure = closure
        super.init(collectionViewLayout: layout)
    }
    
    override func prefersStatusBarHidden() -> Bool {
        return true
    }
    override func viewDidLoad() {
        super.viewDidLoad()

        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false
        
        self.collectionView?.pagingEnabled = true
        self.collectionView?.showsHorizontalScrollIndicator = false
        self.collectionView?.showsVerticalScrollIndicator = false
        self.collectionView?.bounces = false
        // Register cell classes
        self.collectionView!.registerClass(UICollectionViewCell.self, forCellWithReuseIdentifier: reuseIdentifier)

        skipButton.addTarget(self, action: "skipHasClicked", forControlEvents: .TouchUpInside)
        self.view.addSubview(skipButton)
        
        self.view.addSubview(pageControl)
        pageControl.numberOfPages = images.count
        pageControl.currentPage = 0
        pageControl.userInteractionEnabled = false
        pageControl.currentPageIndicatorTintColor = UIColor.greenColor()
        // Do any additional setup after loading the view.
        
        enterButton.addTarget(self, action: "skipHasClicked", forControlEvents: .TouchUpInside)
        enterButton.frame = CGRectMake(0, pageControl.frame.origin.y - 30 - 20, kHJMainScreenWidth, 30)
    }

    func skipHasClicked() {
        skipButton.hidden = true
        enterButton.hidden = true
        self.animation(self.collectionView!, isRotate: true, delegate: self)
        NSUserDefaults.standardUserDefaults().setBool(true, forKey: is_First)
        NSUserDefaults.standardUserDefaults().synchronize()
    }
    
    override func animationDidStop(anim: CAAnimation, finished flag: Bool) {
        if nil != self.skipClosure {
            self.skipClosure!()
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: UICollectionViewDataSource

    override func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
        // #warning Incomplete implementation, return the number of sections
        return 1
    }


    override func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of items
        return images.count
    }

    override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier(reuseIdentifier, forIndexPath: indexPath)
        cell.backgroundColor = UIColor.whiteColor()
    
        if nil == cell.backgroundView {
            let imageView = UIImageView(frame: kHJMainScreenBounds)
            imageView.image = UIImage(contentsOfFile: NSBundle.mainBundle().pathForResource(images[indexPath.item], ofType: nil)!)
            cell.backgroundView = imageView
        } else {
            let imageView = cell.backgroundView as! UIImageView
            imageView.image = UIImage(contentsOfFile: NSBundle.mainBundle().pathForResource(images[indexPath.item], ofType: nil)!)
        }
        
        if indexPath.item == images.count - 1{
            cell.contentView.addSubview(enterButton)
            enterButton.hidden = false
        } else if indexPath.item < images.count  - 2{
            enterButton.hidden = true
        }
        
        return cell
    }

    //MARK:UIScrollViewDelegate
    override func scrollViewDidScroll(scrollView: UIScrollView) {
        let page = (self.collectionView?.contentOffset.x)! / kHJMainScreenWidth
        pageControl.currentPage = Int(page)
    }

    
    func animation(view:UIView, isRotate: Bool, delegate: AnyObject) {
        if isRotate {
            view.transform = CGAffineTransformRotate(view.transform, CGFloat(M_1_PI))
        }
        //缩放
        let duration = 0.5
        let scaleAnimation = CABasicAnimation(keyPath: "transform.scale")
        scaleAnimation.fromValue = NSNumber(double: 1)
        scaleAnimation.toValue = NSNumber(double: 2)
        
        //透明度
        let opacityAnimation = CABasicAnimation(keyPath: "opacity")
        opacityAnimation.fromValue = NSNumber(double: 1)
        opacityAnimation.toValue = NSNumber(double: 0)
        
        //组动画
        let group = CAAnimationGroup()
        group.animations = [scaleAnimation, opacityAnimation]
        group.duration = duration
        group.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
        group.speed = 0.5//慢速
        group.delegate = delegate
        group.autoreverses = false// 防止最后显现,不自动返回
        group.fillMode = kCAFillModeForwards//最终会停在终点处
        //removedOnCompletion:默认为YES,代表动画执行完毕后就从图层上移除,图形会恢复到动画执行前的状态。如果想让图层保持显示动画执行后的状态,那就设置为false, 不会会闪一下
        group.removedOnCompletion = false //这个一定要设置为NO,不然会闪一下
        view.layer.addAnimation(group, forKey: nil)
    }
}
